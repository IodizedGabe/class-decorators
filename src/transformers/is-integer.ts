import { composePropertyDecorators, IsIntegerOptions } from '@hippo-oss/dto-decorators';
import { Type } from 'class-transformer';

import { createBasePropertyDecorators } from './base';

export function IsInteger(options: IsIntegerOptions = {}): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // convert string to number
        Type(() => Number),
    ]);
}
