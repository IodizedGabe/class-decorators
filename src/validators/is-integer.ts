import { composePropertyDecorators, IsIntegerOptions } from '@hippo-oss/dto-decorators';
import { IsInt, Max, Min } from 'class-validator';

import { createBasePropertyDecorators } from './base';

export function IsInteger(options: IsIntegerOptions = {}): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // validate data as an integer
        IsInt({ each: !!options.isArray }),

        // maybe: add a maximum value
        options?.maxValue !== undefined ? Max(options.maxValue, { each: !!options.isArray }) : undefined,

        // maybe: add a minimum value
        options?.minValue !== undefined ? Min(options.minValue, { each: !!options.isArray }) : undefined,
    ]);
}
