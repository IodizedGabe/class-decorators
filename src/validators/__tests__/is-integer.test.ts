import { validateSync } from 'class-validator';

import { IsInteger } from '../is-integer';

describe('IsInteger', () => {
    it('allows optional integer to be omitted', () => {
        class Fixture {
            @IsInteger({
                optional: true,
            })
            property!: number;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('allows nullable integer to be omitted', () => {
        class Fixture {
            @IsInteger({
                nullable: true,
            })
            property!: number;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('does not allow required integer to be omitted', () => {
        class Fixture {
            @IsInteger({})
            property!: number;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow non-integer', () => {
        class Fixture {
            @IsInteger({})
            property!: number;
        }

        const obj = new Fixture();
        obj.property = '42' as unknown as number;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow float', () => {
        class Fixture {
            @IsInteger({})
            property!: number;
        }

        const obj = new Fixture();
        obj.property = 42.1;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow integer that is too small', () => {
        class Fixture {
            @IsInteger({
                minValue: 0,
            })
            property!: number;
        }

        const obj = new Fixture();
        obj.property = -1;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow integer that is too large', () => {
        class Fixture {
            @IsInteger({
                maxValue: 0,
            })
            property!: number;
        }

        const obj = new Fixture();
        obj.property = 1;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('allows integer', () => {
        class Fixture {
            @IsInteger({
                maxValue: 2,
                minValue: 0,
            })
            property!: number;
        }

        const obj = new Fixture();
        obj.property = 1;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
});
