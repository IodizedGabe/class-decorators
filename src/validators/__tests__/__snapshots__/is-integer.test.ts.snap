// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsInteger does not allow float 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isInt": "property must be an integer number",
    },
    "property": "property",
    "target": Fixture {
      "property": 42.1,
    },
    "value": 42.1,
  },
]
`;

exports[`IsInteger does not allow integer that is too large 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "max": "property must not be greater than 0",
    },
    "property": "property",
    "target": Fixture {
      "property": 1,
    },
    "value": 1,
  },
]
`;

exports[`IsInteger does not allow integer that is too small 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "min": "property must not be less than 0",
    },
    "property": "property",
    "target": Fixture {
      "property": -1,
    },
    "value": -1,
  },
]
`;

exports[`IsInteger does not allow non-integer 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isInt": "property must be an integer number",
    },
    "property": "property",
    "target": Fixture {
      "property": "42",
    },
    "value": "42",
  },
]
`;

exports[`IsInteger does not allow required integer to be omitted 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "property should not be null or undefined",
      "isInt": "property must be an integer number",
    },
    "property": "property",
    "target": Fixture {},
    "value": undefined,
  },
]
`;
